Source/CarthageKit/Project.swift

```diff
@@ -413,24 +413,52 @@ public final class Project { // swiftlint:disable:this type_body_length
 			.map(ResolvedCartfile.init)
 	}
 
+	/// Attempts to determine the latest version (whether satisfiable or not)
+	/// of the project's Carthage dependencies.
+	///
+	/// This will fetch dependency repositories as necessary, but will not check
+	/// them out into the project's working directory.
+	private func latestDependencies() -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+		return loadResolvedCartfile()
+			.flatMap(.merge) { cartfile -> SignalProducer<(Dependency, PinnedVersion), CarthageError> in
+				return SignalProducer(cartfile.dependencies.keys)
+					.flatMap(FlattenStrategy.merge) { dependency -> SignalProducer<(Dependency, PinnedVersion), CarthageError> in
+						// For finding the latest version the semantic version is necessary, but it is necessary to preserve the pinned version
+						typealias Version = (semantic: SemanticVersion, pinned: PinnedVersion)
+						let versions = self.versions(for: dependency)
+							.filterMap { version -> Version? in
+								guard let semanticVersion = SemanticVersion.from(version).value else { return nil }
+								return (semantic: semanticVersion, pinned: version)
+							}
+							.collect()
+						let latestVersion = versions
+							.filterMap { $0.max(by: { $0.semantic < $1.semantic }) }
+							.map { $0.pinned }
+						return SignalProducer(value: dependency).combineLatest(with: latestVersion)
```

Comment:
Instead of this, I'd suggest having a single `SignalProducer` chain.

```
typealias Version = (semantic: SemanticVersion, pinned: PinnedVersion)
return self
    .versions(for: dependency)
    .filterMap { version -> Version? in
        ...
    }
    .collect()
    .filterMap { ... }
    .map { (dependency, $0.pinned) }
```